### ISSUE TITLE
Upgrading from v2.x to 4.x ISSUE

### ISSUE THREAD

**Issue by jnystrom** [dmfay/massive-js#459]
OK, so I am getting around to upgrading.  I have upgraded to v4.0.  I have changed the `Massive` connection call to use `deasync` like you suggested here: #412.  

All of a sudden I am not getting TONS of these errors:

```
Creating a duplicate QueryFile object for the same file - 
    /Users/johnnystrom/dev/remedy/coreApi/node_modules/massive/lib/scripts/tables-whitelist.sql
    at loader.queryFiles.files.reduce (/Users/johnnystrom/dev/remedy/coreApi/node_modules/massive/lib/database.js:157:43)
    at filesPromise.$p.then.files (/Users/johnnystrom/dev/remedy/coreApi/node_modules/massive/lib/database.js:156:38)
```

I am using the repository pattern and have multiple repositories that inherit ( using Typescript) from my MassiveRepository class.  This all worked great with v2.X, now I am getting these errors along with:

`error: too many connections for role <db connection name>`

ideas?

### BaseRepo
```
import { IRepository } from './interfaces/IRepository';
import { inject, injectable } from 'inversify';
import Massive = require('massive');
import moment = require('moment');
import humps = require('humps');
import deasync= require('deasync');

@injectable()
export class MassiveRepository implements IRepository {
    protected db: any;
    constructor( @inject('config') private config) {
        this.db = deasync(Massive(config.databaseUrl));
    }

    public findOne(tableName: string, criteria): Promise<any> {
        return new Promise((resolve, reject) => {
            if (criteria) {
                this.db[tableName].findOne(humps.decamelizeKeys(criteria), (err, result: any) => {
                    if (!err) {
                        return resolve(humps.camelizeKeys(result));

                    } else {
                        return reject(err);
                    }
                });
            } else {
                return reject('Criteria is required for using findOne');
            }
        });
    }
```

### Child Repo

```
import { MassiveRepository } from './massiveRepository';
import { IMessageRepository } from './interfaces/IMessageRepository';
import { inject } from 'inversify';
import { ClassifierModel, ClassifierClass, NlpMessage, ScoredMessage, Customer, MessageTag, MessageTagStatus } from '@remedy/models';
import humps = require('humps');
import _ = require('lodash');
export class MessageRepository extends MassiveRepository implements IMessageRepository {

    constructor( @inject('config') config) {
        super(config);
    }
```

**Comment by jnystrom** [comment]
I know a new instance is created for each repo I create.  This was never a problem in v2, I am confused why this would be an issue in the new version.

**Comment by dmfay** [comment]
Here's why it's an issue:

1. QueryFiles are cached at the driver level, and it warns if you reload the same file unnecessarily. You can ignore warnings if you prefer by passing a [driverConfig](https://vitaly-t.github.io/pg-promise/module-pg-promise.html) as the third argument to the Massive constructor.
2. You're connecting with a connection string; in v2, I can't remember whether that created a pool for you. If it didn't, then your connection would have been slower and prone to blocking but you wouldn't have run up against the limit so quickly. Or possibly you just have more repositories now. Either way, this is a Postgres limitation (look into the `max_connections` setting) and a Postgres error message.

Bottom line, Massive is meant to be connected once and for everything to route through a single instance. You can still use the repository pattern, but all your repositories should be using the same Massive object instead of maintaining their own. I don't know if you can create a static property on the parent class with TypeScript, but I'd look into that.

**Comment by vitaly-t** [comment]
> Massive is meant to be connected once and for everything to route through a single instance. 

And not following this will cause all kinds of problems.

**Comment by jnystrom** [comment]
Thanks, I got it all working!
